Some timing tests fail on ARM

## Context & versions
mithril: v2403.1
Platform: ARM64, https://www.solid-run.com/embedded-networking/nxp-lx2160a-family/cex7-lx2160/#carrier-boards

## Steps to reproduce
make test

## Actual behavior
> test utils::progress_reporter::tests::check_seconds_elapsed_in_json_report_with_less_than_100_milliseconds ... FAILED
> test utils::progress_reporter::tests::check_seconds_elapsed_in_json_report_with_more_than_100_milliseconds ... FAILED

> failures:

> ---- utils::progress_reporter::tests::check_seconds_elapsed_in_json_report_with_more_than_100_milliseconds stdout ----
> thread 'utils::progress_reporter::tests::check_seconds_elapsed_in_json_report_with_more_than_100_milliseconds' panicked at mithril-client-cli/src/utils/progress_reporter.rs:160:9:
> Not expected value in json output: {"timestamp": "2024-01-27T00:01:31.427344062+00:00", "bytes_downloaded": 0, "bytes_total": 10, "seconds_left": 0.000, "seconds_elapsed": 5.126}
> note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

> ---- utils::progress_reporter::tests::check_seconds_elapsed_in_json_report_with_less_than_100_milliseconds stdout ----
> thread 'utils::progress_reporter::tests::check_seconds_elapsed_in_json_report_with_less_than_100_milliseconds' panicked at mithril-client-cli/src/utils/progress_reporter.rs:173:9:
> Not expected value in json output: {"timestamp": "2024-01-27T00:01:31.426460563+00:00", "bytes_downloaded": 0, "bytes_total": 10, "seconds_left": 0.000, "seconds_elapsed": 5.007}


> failures:
>     utils::progress_reporter::tests::check_seconds_elapsed_in_json_report_with_less_than_100_milliseconds
>     utils::progress_reporter::tests::check_seconds_elapsed_in_json_report_with_more_than_100_milliseconds

> test result: FAILED. 18 passed; 2 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.14s


Is this platform is too slow?

On an Ampere system:

* Mithril Aggregator
  * passes `make test`
  * fails `make debug` with `Error: configuration deserialize error` 
* Mithril Signer
  * passes `make test`
  * fails `make debug` with `Error: configuration deserialize error`
* Mithril Client
  * fails `make test` with:

```
failures:

---- utils::cardano_db_download_checker::test::unix_only::return_error_if_directory_could_not_be_created stdout ----
thread 'utils::cardano_db_download_checker::test::unix_only::return_error_if_directory_could_not_be_created' panicked at mithril-client-cli/src/utils/cardano_db_download_checker.rs:253:18:
ensure_dir_exist should fail: ()
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

---- utils::cardano_db_download_checker::test::unix_only::return_error_if_existing_directory_is_not_writable stdout ----
thread 'utils::cardano_db_download_checker::test::unix_only::return_error_if_existing_directory_is_not_writable' panicked at mithril-client-cli/src/utils/cardano_db_download_checker.rs:282:14:
check_prerequisites should fail: ()


failures:
    utils::cardano_db_download_checker::test::unix_only::return_error_if_directory_could_not_be_created
    utils::cardano_db_download_checker::test::unix_only::return_error_if_existing_directory_is_not_writable

test result: FAILED. 22 passed; 2 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.18s

error: test failed, to rerun pass `--lib`
make: *** [Makefile:23: test] Error 101
```

--- a/mithril-client-cli/Cargo.toml
+++ b/mithril-client-cli/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "mithril-client-cli"
-version = "0.8.1"
+version = "0.8.2"
 description = "A Mithril Client"
 authors = { workspace = true }
 edition = { workspace = true }

--- a/mithril-client-cli/src/utils/progress_reporter.rs
+++ b/mithril-client-cli/src/utils/progress_reporter.rs
@@ -237,8 +237,8 @@ mod tests {
     fn check_seconds_left_and_elapsed_time_are_used_by_the_formatter() {
         // 4 steps
         let progress_bar = ProgressBar::new(4);
-        // 1 step done in 15 ms, left 45ms to finish the 4th steps
-        sleep(Duration::from_millis(15));
+        // 1 step done in 150 ms, left 450ms to finish the 4th steps
+        sleep(Duration::from_millis(150));
         progress_bar.set_position(1);
 
         let json_string = ProgressBarJsonFormatter::format(&progress_bar);
@@ -247,19 +247,19 @@ mod tests {
         let milliseconds_elapsed = progress_bar.elapsed().as_millis();
 
         // Milliseconds in json may not be exactly the same as the one we get because of the test duration.
-        // We need to have a difference not more than 4ms to keep the same 2 first milliseconds digits.
-        let delta = 4;
+        // We need to have a difference not more than 49ms to keep the same 1 first milliseconds digits.
+        let delta = 49;
 
-        assert!(((45 - delta)..=(45 + delta)).contains(&milliseconds_left));
+        assert!(((450 - delta)..=(450 + delta)).contains(&milliseconds_left));
         assert!(
-            json_string.contains(r#""seconds_left": 0.04"#), // Should be close to 0.045
+            json_string.contains(r#""seconds_left": 0.4"#), // Should be close to 0.450
             "Not expected value in json output: {}",
             json_string
         );
 
-        assert!(((15 - delta)..(15 + delta)).contains(&milliseconds_elapsed));
+        assert!(((150 - delta)..(150 + delta)).contains(&milliseconds_elapsed));
         assert!(
-            json_string.contains(r#""seconds_elapsed": 0.01"#), // Should be close to 0.015
+            json_string.contains(r#""seconds_elapsed": 0.1"#), // Should be close to 0.150
             "Not expected value in json output: {}",
             json_string
         );

--- a/mithril-client-cli/src/utils/progress_reporter.rs
+++ b/mithril-client-cli/src/utils/progress_reporter.rs
@@ -166,6 +166,7 @@ mod tests {
 
     use super::*;
     use indicatif::ProgressBar;
+    use serde_json::Value;
 
     #[test]
     fn check_seconds_formatting_in_json_report_with_more_than_100_milliseconds() {
@@ -235,54 +236,60 @@ mod tests {
 
     #[test]
     fn check_seconds_left_and_elapsed_time_are_used_by_the_formatter() {
-        let expected_milliseconds_left: u128 = 45;
-        let expected_milliseconds_elapsed: u128 = 15;
+        fn format_duration(duration: &Duration) -> String {
+            format!("{}.{}", duration.as_secs(), duration.subsec_nanos())
+        }
+        fn round_at_ms(duration: Duration) -> Duration {
+            Duration::from_millis(duration.as_millis() as u64)
+        }
 
         // 4 steps
         let progress_bar = ProgressBar::new(4);
-        // 1 step done in 150 ms, left 450ms to finish the 4th steps
-        sleep(Duration::from_millis(expected_milliseconds_elapsed as u64));
+        // 1 step done in 15 ms, left 45ms to finish the 4th steps
+        sleep(Duration::from_millis(15));
         progress_bar.set_position(1);
+
+        let duration_left_before = round_at_ms(progress_bar.eta());
+        let duration_elapsed_before = round_at_ms(progress_bar.elapsed());
 
         let json_string = ProgressBarJsonFormatter::format(&progress_bar);
 
-        let milliseconds_left = progress_bar.eta().as_millis();
-        let milliseconds_elapsed = progress_bar.elapsed().as_millis();
+        let duration_left_after = round_at_ms(progress_bar.eta());
+        let duration_elapsed_after = round_at_ms(progress_bar.elapsed());
 
         // Milliseconds in json may not be exactly the same as the one we get because of the test duration.
-        // We need to have a difference not more than 49ms to keep the same 1 first milliseconds digits.
-        let delta = 4;
-        // TODO: to remove, it-s just to investigation on CI.
-        let error_message = format!(
-            "\n- milliseconds_left:{milliseconds_left}\n- milliseconds_elapsed:{milliseconds_elapsed}\n- json_string:{json_string}"
+        let delta = 0.1;
+
+        let json_value: Value = serde_json::from_str(&json_string).unwrap();
+        let seconds_left = json_value["seconds_left"].as_f64().unwrap();
+        let seconds_elapsed = json_value["seconds_elapsed"].as_f64().unwrap();
+
+        // We check that we pass the right values to format checking that time left is 3 times the time elapsed
+        assert!(
+            seconds_elapsed * 3.0 - delta < seconds_left
+                && seconds_left < seconds_elapsed * 3.0 + delta,
+            "seconds_left should be close to 3*{} but it's {}.",
+            &seconds_elapsed,
+            &seconds_left
         );
 
+        let duration_left = Duration::from_secs_f64(seconds_left);
         assert!(
-            ((expected_milliseconds_left - delta)..=(expected_milliseconds_left + delta))
-                .contains(&milliseconds_left),
-            "milliseconds_left should be close to {} but it's {}. {}",
-            &expected_milliseconds_left,
-            &milliseconds_left,
-            &error_message
-        );
-        assert!(
-            json_string.contains(r#""seconds_left": 0.04"#), // Should be close to 0.450
-            "Not expected value in json output: {}",
-            json_string
+            duration_left_before <= duration_left && duration_left <= duration_left_after,
+            "Duration left: {} should be between {} and {}",
+            format_duration(&duration_left),
+            format_duration(&duration_left_before),
+            format_duration(&duration_left_after),
         );
 
+        let duration_elapsed = Duration::from_secs_f64(seconds_elapsed);
         assert!(
-            ((expected_milliseconds_elapsed - delta)..=(expected_milliseconds_elapsed + delta))
-                .contains(&milliseconds_elapsed),
-            "milliseconds_elapsed should be close to {} but it's {}. {}",
-            &expected_milliseconds_elapsed,
-            &milliseconds_elapsed,
-            &error_message
-        );
-        assert!(
-            json_string.contains(r#""seconds_elapsed": 0.01"#), // Should be close to 0.150
-            "Not expected value in json output: {}",
-            json_string
+            duration_elapsed_before <= duration_elapsed
+                && duration_elapsed <= duration_elapsed_after,
+            "Duration elapsed: {} should be between {} and {}",
+            format_duration(&duration_elapsed),
+            format_duration(&duration_elapsed_before),
+            format_duration(&duration_elapsed_after),
         );
     }
 }

--- a/mithril-client-cli/src/utils/progress_reporter.rs
+++ b/mithril-client-cli/src/utils/progress_reporter.rs
@@ -78,15 +78,31 @@ pub struct ProgressBarJsonFormatter;
 impl ProgressBarJsonFormatter {
     /// Get a json formatted string given the progress bar status
     pub fn format(progress_bar: &ProgressBar) -> String {
-        format!(
-            r#"{{"timestamp": "{}", "bytes_downloaded": {}, "bytes_total": {}, "seconds_left": {}.{:0>3}, "seconds_elapsed": {}.{:0>3}}}"#,
+        ProgressBarJsonFormatter::format_values(
             Utc::now().to_rfc3339(),
             progress_bar.position(),
             progress_bar.length().unwrap_or(0),
-            progress_bar.eta().as_secs(),
-            progress_bar.eta().subsec_millis(),
-            progress_bar.elapsed().as_secs(),
-            progress_bar.elapsed().subsec_millis(),
+            progress_bar.eta(),
+            progress_bar.elapsed(),
+        )
+    }
+
+    fn format_values(
+        timestamp: String,
+        bytes_downloaded: u64,
+        bytes_total: u64,
+        duration_left: Duration,
+        duration_elapsed: Duration,
+    ) -> String {
+        format!(
+            r#"{{"timestamp": "{}", "bytes_downloaded": {}, "bytes_total": {}, "seconds_left": {}.{:0>3}, "seconds_elapsed": {}.{:0>3}}}"#,
+            timestamp,
+            bytes_downloaded,
+            bytes_total,
+            duration_left.as_secs(),
+            duration_left.subsec_millis(),
+            duration_elapsed.as_secs(),
+            duration_elapsed.subsec_millis(),
         )
     }
 }
@@ -152,11 +168,20 @@ mod tests {
     use indicatif::ProgressBar;
 
     #[test]
-    fn check_seconds_elapsed_in_json_report_with_more_than_100_milliseconds() {
-        let progress_bar = ProgressBar::new(10).with_elapsed(Duration::from_millis(5124));
+    fn check_seconds_formatting_in_json_report_with_more_than_100_milliseconds() {
+        let json_string = ProgressBarJsonFormatter::format_values(
+            "".to_string(),
+            0,
+            0,
+            Duration::from_millis(7569),
+            Duration::from_millis(5124),
+        );
 
-        let json_string = ProgressBarJsonFormatter::format(&progress_bar);
-
+        assert!(
+            json_string.contains(r#""seconds_left": 7.569"#),
+            "Not expected value in json output: {}",
+            json_string
+        );
         assert!(
             json_string.contains(r#""seconds_elapsed": 5.124"#),
             "Not expected value in json output: {}",
@@ -165,11 +190,20 @@ mod tests {
     }
 
     #[test]
-    fn check_seconds_elapsed_in_json_report_with_less_than_100_milliseconds() {
-        let progress_bar = ProgressBar::new(10).with_elapsed(Duration::from_millis(5004));
+    fn check_seconds_formatting_in_json_report_with_less_than_100_milliseconds() {
+        let json_string = ProgressBarJsonFormatter::format_values(
+            "".to_string(),
+            0,
+            0,
+            Duration::from_millis(7006),
+            Duration::from_millis(5004),
+        );
 
-        let json_string = ProgressBarJsonFormatter::format(&progress_bar);
-
+        assert!(
+            json_string.contains(r#""seconds_left": 7.006"#),
+            "Not expected value in json output: {}",
+            json_string
+        );
         assert!(
             json_string.contains(r#""seconds_elapsed": 5.004"#),
             "Not expected value in json output: {}",
@@ -178,38 +212,54 @@ mod tests {
     }
 
     #[test]
-    fn check_seconds_left_in_json_report_with_more_than_100_milliseconds() {
-        let half_position = 5;
-        let progress_bar = ProgressBar::new(half_position * 2);
-        sleep(Duration::from_millis(123));
-        progress_bar.set_position(half_position);
-        let json_string = ProgressBarJsonFormatter::format(&progress_bar);
+    fn check_seconds_formatting_in_json_report_with_milliseconds_ending_by_zeros() {
+        let json_string = ProgressBarJsonFormatter::format_values(
+            "".to_string(),
+            0,
+            0,
+            Duration::from_millis(7200),
+            Duration::from_millis(5100),
+        );
 
-        let milliseconds = progress_bar.eta().subsec_millis();
-        assert!(milliseconds > 100);
         assert!(
-            json_string.contains(&format!(r#""seconds_left": 0.{}"#, milliseconds)),
+            json_string.contains(r#""seconds_left": 7.200"#),
+            "Not expected value in json output: {}",
+            json_string
+        );
+        assert!(
+            json_string.contains(r#""seconds_elapsed": 5.100"#),
             "Not expected value in json output: {}",
             json_string
         );
     }
 
     #[test]
-    fn check_seconds_left_in_json_report_with_less_than_100_milliseconds() {
-        let half_position = 5;
-        let progress_bar = ProgressBar::new(half_position * 2);
-        sleep(Duration::from_millis(1));
-        progress_bar.set_position(half_position);
+    fn check_seconds_left_and_elapsed_time_are_used_by_the_formatter() {
+        // 4 steps
+        let progress_bar = ProgressBar::new(4);
+        // 1 step done in 15 ms, left 45ms to finish the 4th steps
+        sleep(Duration::from_millis(15));
+        progress_bar.set_position(1);
+
         let json_string = ProgressBarJsonFormatter::format(&progress_bar);
 
+        let milliseconds_left = progress_bar.eta().as_millis();
+        let milliseconds_elapsed = progress_bar.elapsed().as_millis();
+
+        // Milliseconds in json may not be exactly the same as the one we get because of the test duration.
+        // We need to have a difference not more than 4ms to keep the same 2 first milliseconds digits.
+        let delta = 4;
+
+        assert!(((45 - delta)..=(45 + delta)).contains(&milliseconds_left));
         assert!(
-            json_string.contains(r#""seconds_left": 0.0"#),
+            json_string.contains(r#""seconds_left": 0.04"#), // Should be close to 0.045
             "Not expected value in json output: {}",
             json_string
         );
 
+        assert!(((15 - delta)..(15 + delta)).contains(&milliseconds_elapsed));
         assert!(
-            !json_string.contains(r#""seconds_left": 0.000"#),
+            json_string.contains(r#""seconds_elapsed": 0.01"#), // Should be close to 0.015
             "Not expected value in json output: {}",
             json_string
         );

--- a/mithril-client-cli/src/utils/progress_reporter.rs
+++ b/mithril-client-cli/src/utils/progress_reporter.rs
@@ -235,10 +235,13 @@ mod tests {
 
     #[test]
     fn check_seconds_left_and_elapsed_time_are_used_by_the_formatter() {
+        let expected_milliseconds_left: u128 = 450;
+        let expected_milliseconds_elapsed: u128 = 150;
+
         // 4 steps
         let progress_bar = ProgressBar::new(4);
         // 1 step done in 150 ms, left 450ms to finish the 4th steps
-        sleep(Duration::from_millis(150));
+        sleep(Duration::from_millis(expected_milliseconds_elapsed as u64));
         progress_bar.set_position(1);
 
         let json_string = ProgressBarJsonFormatter::format(&progress_bar);
@@ -250,14 +253,26 @@ mod tests {
         // We need to have a difference not more than 49ms to keep the same 1 first milliseconds digits.
         let delta = 49;
 
-        assert!(((450 - delta)..=(450 + delta)).contains(&milliseconds_left));
+        assert!(
+            ((expected_milliseconds_left - delta)..=(expected_milliseconds_left + delta))
+                .contains(&milliseconds_left),
+            "milliseconds_left should be close to {} but it's {}",
+            &expected_milliseconds_left,
+            &milliseconds_left
+        );
         assert!(
             json_string.contains(r#""seconds_left": 0.4"#), // Should be close to 0.450
             "Not expected value in json output: {}",
             json_string
         );
 
-        assert!(((150 - delta)..(150 + delta)).contains(&milliseconds_elapsed));
+        assert!(
+            ((expected_milliseconds_elapsed - delta)..=(expected_milliseconds_elapsed + delta))
+                .contains(&milliseconds_elapsed),
+            "milliseconds_elapsed should be close to {} but it's {}",
+            &expected_milliseconds_elapsed,
+            &milliseconds_elapsed
+        );
         assert!(
             json_string.contains(r#""seconds_elapsed": 0.1"#), // Should be close to 0.150
             "Not expected value in json output: {}",

--- a/mithril-client-cli/src/utils/progress_reporter.rs
+++ b/mithril-client-cli/src/utils/progress_reporter.rs
@@ -235,8 +235,8 @@ mod tests {
 
     #[test]
     fn check_seconds_left_and_elapsed_time_are_used_by_the_formatter() {
-        let expected_milliseconds_left: u128 = 450;
-        let expected_milliseconds_elapsed: u128 = 150;
+        let expected_milliseconds_left: u128 = 45;
+        let expected_milliseconds_elapsed: u128 = 15;
 
         // 4 steps
         let progress_bar = ProgressBar::new(4);
@@ -251,7 +251,7 @@ mod tests {
 
         // Milliseconds in json may not be exactly the same as the one we get because of the test duration.
         // We need to have a difference not more than 49ms to keep the same 1 first milliseconds digits.
-        let delta = 49;
+        let delta = 4;
         // TODO: to remove, it-s just to investigation on CI.
         let error_message = format!(
             "\n- milliseconds_left:{milliseconds_left}\n- milliseconds_elapsed:{milliseconds_elapsed}\n- json_string:{json_string}"
@@ -266,7 +266,7 @@ mod tests {
             &error_message
         );
         assert!(
-            json_string.contains(r#""seconds_left": 0.4"#), // Should be close to 0.450
+            json_string.contains(r#""seconds_left": 0.04"#), // Should be close to 0.450
             "Not expected value in json output: {}",
             json_string
         );
@@ -280,7 +280,7 @@ mod tests {
             &error_message
         );
         assert!(
-            json_string.contains(r#""seconds_elapsed": 0.1"#), // Should be close to 0.150
+            json_string.contains(r#""seconds_elapsed": 0.01"#), // Should be close to 0.150
             "Not expected value in json output: {}",
             json_string
         );

--- a/mithril-client-cli/src/utils/progress_reporter.rs
+++ b/mithril-client-cli/src/utils/progress_reporter.rs
@@ -252,13 +252,18 @@ mod tests {
         // Milliseconds in json may not be exactly the same as the one we get because of the test duration.
         // We need to have a difference not more than 49ms to keep the same 1 first milliseconds digits.
         let delta = 49;
+        // TODO: to remove, it-s just to investigation on CI.
+        let error_message = format!(
+            "\n- milliseconds_left:{milliseconds_left}\n- milliseconds_elapsed:{milliseconds_elapsed}\n- json_string:{json_string}"
+        );
 
         assert!(
             ((expected_milliseconds_left - delta)..=(expected_milliseconds_left + delta))
                 .contains(&milliseconds_left),
-            "milliseconds_left should be close to {} but it's {}",
+            "milliseconds_left should be close to {} but it's {}. {}",
             &expected_milliseconds_left,
-            &milliseconds_left
+            &milliseconds_left,
+            &error_message
         );
         assert!(
             json_string.contains(r#""seconds_left": 0.4"#), // Should be close to 0.450
@@ -269,9 +274,10 @@ mod tests {
         assert!(
             ((expected_milliseconds_elapsed - delta)..=(expected_milliseconds_elapsed + delta))
                 .contains(&milliseconds_elapsed),
-            "milliseconds_elapsed should be close to {} but it's {}",
+            "milliseconds_elapsed should be close to {} but it's {}. {}",
             &expected_milliseconds_elapsed,
-            &milliseconds_elapsed
+            &milliseconds_elapsed,
+            &error_message
         );
         assert!(
             json_string.contains(r#""seconds_elapsed": 0.1"#), // Should be close to 0.150

